### ISSUE TITLE
DPO Loss kernel is too slow

### ISSUE THREAD

**Issue by Manan17** [linkedin/Liger-Kernel#768]
### 🐛 Describe the bug

I tried running DPO Loss with TRL's DPOTrainer on a single H100 GPU, 
The estimated time to complete with Liger loss was 9 hours and without Liger was 2 hours 30 minutes. 
(I set use_liger_loss=True)

### Reproduce

```
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
train_dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")

training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO", logging_steps=10, use_liger_loss=True)
trainer = DPOTrainer(model=model, args=training_args, processing_class=tokenizer, train_dataset=train_dataset)
trainer.train()
```

### Versions

Operating System: Linux-5.15.180.1-1.cm2-x86_64-with-glibc2.35
Python version: 3.10.14
Liger Kernel version: 0.5.10
PyTorch version: 2.7.1+cu126
CUDA version: 12.6
Triton version: 3.3.1
Transformers version: 4.52.4

**Comment by Tcc0403** [comment]
Is chunked loss dpo slower than normal dpo in benchmarking? If not, we should check whether recompiling happened in real use case.

**Comment by Manan17** [comment]
Yes, the loss dpo is slower than the normal dpo in benchmarking

**Comment by Manan17** [comment]
@Tcc0403 you can see this torch profiling I did:
Liger's DPO takes 4ms, while huggingface dpo takes 292 µs
<img width="904" height="659" alt="Image" src="https://github.com/user-attachments/assets/d7ad4b48-f031-4da7-a69c-1c07dd3ed583" />
<img width="507" height="650" alt="Image" src="https://github.com/user-attachments/assets/3ef517ac-ff40-4db4-9c6d-45f36c5df3d1" />

**Comment by Tcc0403** [comment]
@Manan17 
Half of the time spent on torch.compile. I wonder what causes this amount of overhead. Can you share the script you use?

**Comment by Manan17** [comment]
Just pass `use_liger_loss=True` in the DPOConfig for running it with liger.
```
# train_dpo_profiled.py
import torch
from torch.profiler import profile, record_function, ProfilerActivity
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer
import os

def main():
    print("Setting up profiler...")
    
    # Clean up old logs
    import shutil
    if os.path.exists('./tb_logs'):
        shutil.rmtree('./tb_logs')
    
    # Create profiler with proper TensorBoard integration
    def trace_handler(p):
        output = p.key_averages().table(sort_by="self_cuda_time_total" if torch.cuda.is_available() else "self_cpu_time_total", row_limit=10)
        print(f"Step {p.step_num}: Top operations\n{output}")
        p.export_chrome_trace(f"./tb_logs/trace_{p.step_num}.json")
    
    prof = profile(
        activities=[
            ProfilerActivity.CPU,
            ProfilerActivity.CUDA if torch.cuda.is_available() else None,
        ],
        schedule=torch.profiler.schedule(
            wait=1,
            warmup=1, 
            active=2,
            repeat=1
        ),
        on_trace_ready=trace_handler,
        record_shapes=True,
        profile_memory=True,
        with_stack=True,
    )
    
    # Make sure output directory exists
    os.makedirs('./tb_logs', exist_ok=True)
    
    prof.start()
    
    try:
        # Load components
        with record_function("model_loading"):
            print("Loading model and tokenizer...")
            model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
            tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
        
        with record_function("dataset_loading"):
            print("Loading dataset...")
            train_dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")
        
        with record_function("trainer_setup"):
            print("Setting up trainer...")
            training_args = DPOConfig(
                output_dir="Qwen2-0.5B-DPO",
                max_steps=6,
                per_device_train_batch_size=1,
                logging_steps=1,
                save_steps=1000,
                dataloader_num_workers=0,
            )
            
            trainer = DPOTrainer(
                model=model, 
                args=training_args, 
                processing_class=tokenizer, 
                train_dataset=train_dataset
            )
        
        print("Starting profiled training...")
        
        # Manual training loop with profiler steps
        trainer.model.train()
        train_dataloader = trainer.get_train_dataloader()
        
        step = 0
        for batch in train_dataloader:
            if step >= 6:  # Match max_steps
                break
                
            with record_function(f"training_step_{step}"):
                # Perform training step
                trainer.training_step(trainer.model, batch)
            
            prof.step()  # Important: step the profiler
            step += 1
            print(f"Completed step {step}")
        
    finally:
        prof.stop()
    
    print("\n" + "="*60)
    print("PROFILING COMPLETE")
    print("="*60)
    
    # Generate final summary
    try:
        print("\nOverall Summary:")
        key_averages = prof.key_averages()
        
        if torch.cuda.is_available():
            print("\nTop CUDA operations:")
            print(key_averages.table(sort_by="cuda_time_total", row_limit=10))
        
        print("\nTop CPU operations:")
        print(key_averages.table(sort_by="cpu_time_total", row_limit=10))
        
        print("\nMemory usage:")
        print(key_averages.table(sort_by="cpu_memory_usage", row_limit=10))
        
        # Export final trace
        prof.export_chrome_trace("./tb_logs/full_trace.json")
        
        print(f"\nFiles created:")
        print(f"- Chrome traces: ./tb_logs/trace_*.json")
        print(f"- Full trace: ./tb_logs/full_trace.json")
        print(f"\nTo view traces:")
        print(f"1. Open Chrome and go to chrome://tracing")
        print(f"2. Load any of the trace_*.json files")
        
    except Exception as e:
        print(f"Error in summary: {e}")

if __name__ == "__main__":
    main()
```

**Comment by Tcc0403** [comment]
@Manan17 Thanks for the script.

I think this part of python overhead is overlap with gpu computations prior to loss. Although the liger dpo loss itself is much slower the original dpo loss, the overhead should be neglectable since it doesn't block the gpu computation.


Can you check the gpu stream with your testing gpu if the torch.compile overhead is fully overlapped?


## ~~Profiling~~
~~**Note that I'm running with a consumer gpu. The computation might be much shorter on data center gpus so that it couldn't hide the torch.compile overhead.**~~
Edit: my profiling shouldn't be referenced, it's extremely inaccurate.

## Summary
From above figures, we can confirm that the overhead doesn't make computations slower, and it is faster if there's only one chunk. Thus, I think it's not the main bottleneck in trainings. 

By the way, with liger dpo loss, the second matmul (ref_model) is computed in fp32 while the orginal is in bf16.

**Comment by Tcc0403** [comment]
Also, I just check the benchmarking. The `forward()` of `DPOLossFunction` is dominated by other overhead, probably instantiation.

https://github.com/linkedin/Liger-Kernel/blob/d70a26b66a40ebdf28e68d0ad421dc9fb20f9e87/benchmarks/08906d3/benchmark.csv#L158-L167

I had addresed this issue in #643 and also posted the correct result with proper measurement. I totally forgot this issue and closed the PR. We should fix it as well.

**Comment by Tcc0403** [comment]
[gist](https://gist.github.com/Tcc0403/ac01f3d089448e8ac83e415594896432)

Profiling trl dpo trainer with and without liger loss on modal with above script
```
modal run modal_dpo_trace.py  
```


## Profiling
The main difference between liger dpo and non liger dpo is about lm_head matmul:
### Original

<img width="1007" height="245" alt="Image" src="https://github.com/user-attachments/assets/cd940e5d-ac1a-4842-a98e-09b68cce64ad" />

4 matmuls in total

### Liger

<img width="1102" height="243" alt="Image" src="https://github.com/user-attachments/assets/6957c9f3-878b-4193-818a-3b115c221d03" />

( 1 for logits + 1 for ref_logits + 2 for grad_input and grad_weight ) * ( batch_size / chunk_size ) = 4 * 4 
= 16 matmuls in total

### Observations

The total number of matmuls for original dpo and liger dpo is as expected, so it wouldn't surprise me if liger dpo takes longer on training.

However, we can see that the duration of the second kernel(`nvjet_tst_256x160_64x4_2x1_v_bz_coopA_TNT`), which is ref_model's lm_head forward, in original dpo training is significantly shorter than others. 

I checked trl [code](https://github.com/huggingface/trl/blob/0462dd7f1230388f98c62990778f1d4db5aedf2f/trl/trainer/dpo_trainer.py#L902) but I couldn't figure out what causes ref_model forward launching another gpu kernel.

**Comment by Manan17** [comment]
Whenever we use liger's dpo loss, it does matmul calculation in fp32 rather than bf16 (done by huggingface trl by default). This was the reason of the slowness.

You can use:
`model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct", torch_dtype=torch.bfloat16)`

So now the speed of liger is little better than of huggingface.
